src/vs/platform/actions/common/actions.ts

```diff
@@ -272,6 +272,7 @@ export interface IMenuChangeEvent {
 export interface IMenu extends IDisposable {
 	readonly onDidChange: Event<IMenuChangeEvent>;
 	getActions(options?: IMenuActionOptions): [string, Array<MenuItemAction | SubmenuItemAction>][];
+	contexts(): ReadonlySet<string>;
```

Comment:
I like this but I would suggest to move it out of `IMenu` and into `IMenuService`, like the following

```ts
export interface IMenuService {
    // ... existing stuff

  getMenuContexts(id: MenuId): ReadonlySet<string>;
}
```

The implementation would be like yours: create, read and dispose the actual menu. We could then complement this with `IMenuService#getMenuActions(id: MenudId, options?: IMenuActionOptions): [string, Array<MenuItemAction | SubmenuItemAction>][];)` which would be the equivalent of: create menu, get actions, dispose menu. 

Those two methods would be the "dead" counter-part of `IMenu` (which is listening and emitting). We could also add `IMenuService#getMenuData` which would encapsulate those two concepts: context and action groups but no eventing. So, maybe alltogether


```ts
interface IMenuData {
  contexts(): ReadonlySet<string>;
  actions: [string, Array<MenuItemAction | SubmenuItemAction>][];
}

export interface IMenuService {
    // ... existing stuff

  getMenuData(id: MenuId): IMenuData;
}
```

IMO this would encapsulate things nicer and clearly separate live IMenu-instances from just menu data which you to handle yourselves.

Reply:
Love it. I'll make the change.

Reply:
Looking at this further, I'm not sure we should separate the contexts from the menu. At the time menu is created it has a specific set of contexts. If you create that menu again later with a different scoped context key service you could have different contexts and different actions. I would propose that we have the `IMenuData`, and then have 2 additions to `IMenuService`:
```ts
/**
 * The "dead" counterpart to `createMenu`.
 * For other components to use.
 **/
getMenuActions(id: MenudId, contextKeyService: IContextKeyService, options?: IMenuActionOptions): [string, Array<MenuItemAction | SubmenuItemAction>][];

/**
 * This is what the tree view will use to create a "dead" menu and get the contexts
 **/
getMenuData(id: MenudId, contextKeyService: IContextKeyService, options?: IMenuActionOptions): IMenuData;
```

Reply:
> At the time menu is created it has a specific set of contexts. If you create that menu again later with a different scoped context key service you could have different contexts and different actions.

Not contexts but context evaluations. The former are statically defined when you register menus items (mostly via Action2) and those (keys) should be enough to know what to listen on. The menu does exactly [that](https://github.com/microsoft/vscode/blob/82104a3a6cf8713b81e5bbd97960dbf5c82a816a/src/vs/platform/actions/common/menuService.ts#L387) which in convenient but wasteful when you have very many menu instances for the same menu-id. 

So, likely we need `getMenuActions` with the signature you have outlined above and then `getMenuData` is without the CKS-argument but just returning the context keys (for listening).

Reply:
Hmm. I think I'm missing something still then since I think we still need the CKS for `getMenuData`. The items that get added to the menu depend on the CKS that is passed in. In the tree, we even make an overlay:

https://github.com/microsoft/vscode/blob/31493f3a024372180610abfa199de4f324712900/src/vs/workbench/browser/parts/views/treeView.ts#L1701-L1704